pkg/eventemitter/eventemitter.go

```diff
@@ -0,0 +1,247 @@
+type EventEmitter struct {
+	client.Client
+	record.EventRecorder
+	eventHandlersCache      map[string]EventDataHandler
+	eventHandlersCachesLock *sync.RWMutex
+	eventLoopContexts       *sync.Map
+}
+
+// EventData will save all event info and handler info for retry.
+type EventData struct {
+	namespace  string
+	objectName string
+	eventtype  string
+	reason     string
+	message    string
+	time       time.Time
+	handlerKey string
+	retryTimes int
+	err        error
+}
+
+type EventDataHandler interface {
+	EmitEvent(eventData EventData, failureFunc func(eventData EventData, err error))
+	CloseHandler()
+}
+
+type EmitData struct {
+	Reason  string `json:"reason"`
+	Message string `json:"message"`
+}
+
+const (
+	CloudEventHTTP = "CloudEventHTTP"
+)
+
+func NewEventEmitter(client client.Client, recorder record.EventRecorder) *EventEmitter {
+	return &EventEmitter{
+		Client:                  client,
+		EventRecorder:           recorder,
+		eventHandlersCache:      map[string]EventDataHandler{},
+		eventHandlersCachesLock: &sync.RWMutex{},
+		eventLoopContexts:       &sync.Map{},
+	}
+}
+
+func (e *EventEmitter) HandleCloudEvents(ctx context.Context, cloudEvent *kedav1alpha1.CloudEvent) error {
+	e.createEventHandlers(ctx, cloudEvent)
+
+	key := cloudEvent.GenerateIdentifier()
+	ctx, cancel := context.WithCancel(ctx)
+
+	// cancel the outdated EventLoop for the same CloudEvent (if exists)
+	value, loaded := e.eventLoopContexts.LoadOrStore(key, cancel)
+	if loaded {
+		cancelValue, ok := value.(context.CancelFunc)
+		if ok {
+			cancelValue()
+		}
+		e.eventLoopContexts.Store(key, cancel)
+	}
+
+	// passing deep copy of CloudEvents to the eventLoop go routines, it's a precaution to not have global objects shared between threads
+	log.V(1).Info("Start CloudEvent loop.")
+	go e.startEventLoop(ctx)
+	return nil
+}
+
+func (e *EventEmitter) DeleteCloudEvents(cloudEvent *kedav1alpha1.CloudEvent) error {
+	key := cloudEvent.GenerateIdentifier()
+	result, ok := e.eventLoopContexts.Load(key)
+	if ok {
+		cancel, ok := result.(context.CancelFunc)
+		if ok {
+			cancel()
+		}
+		e.eventLoopContexts.Delete(key)
+		e.clearEventHandlersCache(cloudEvent)
+	} else {
+		log.V(1).Info("CloudEvent was not found in controller cache", "key", key)
+	}
+
+	return nil
+}
+
+func (e *EventEmitter) createEventHandlers(ctx context.Context, cloudEvents *kedav1alpha1.CloudEvent) {
+	e.eventHandlersCachesLock.Lock()
+	defer e.eventHandlersCachesLock.Unlock()
+
+	key := cloudEvents.GenerateIdentifier()
+
+	clusterName := cloudEvents.Spec.ClusterName
+	if clusterName == "" {
+		clusterName = "default"
+	}
+
+	if cloudEvents.Spec.Destination.CloudEventHTTP != nil {
+		var err error
+		var eventHandler EventDataHandler
+		eventHandler, err = NewCloudEventHTTPHandler(ctx, clusterName, cloudEvents.Spec.Destination.CloudEventHTTP.Uri, log)
+
+		if err != nil {
+			log.Error(err, "create cloudevent handler failed")
+		}
+		e.eventHandlersCache[key+CloudEventHTTP] = eventHandler
+	}
```

Comment:
Maybe we should add an error message here if no handler is selected, it could be more clear (or maybe is checked at other point and I haven't seen it yet)

Reply:
Error or warning?

Reply:
Added a handler creation checking in HandleCloudEvents()